File: frontend/generate-sitemap.py
#!/usr/bin/env python3
import os
import sys
from datetime import datetime

LOCALES = [
    ("en", "en"),
    ("tr", "tr"),
    ("de", "de"),
    ("fr", "fr"),
    ("es", "es"),
    ("ar", "ar"),
    ("ja", "ja"),
    ("zh", "zh"),
    ("ru", "ru"),
    ("ko", "ko"),
]

def add_url(loc, lastmod, changefreq, priority, hreflangs=None):
    url = "  <url>\n    <loc>" + loc + "</loc>\n"
    url += "    <lastmod>" + lastmod + "</lastmod>\n"
    url += "    <changefreq>" + changefreq + "</changefreq>\n"
    url += "    <priority>" + str(priority) + "</priority>"
    if hreflangs:
        for hrf, href in hreflangs:
            url += "\n    <xhtml:link rel=\"alternate\" hreflang=\"" + hrf + "\" href=\"" + href + "\" />"
    url += "\n  </url>"
    return url

def main():
    LOCALES = [
        ("en", "en"),
        ("tr", "tr"),
        ("de", "de"),
        ("fr", "fr"),
        ("es", "es"),
        ("ar", "ar"),
        ("ja", "ja"),
        ("zh", "zh"),
        ("ru", "ru"),
        ("ko", "ko"),
    ]

    def hreflang_links(path):
        links = []
        for lc, hrf in LOCALES:
            prefix = "/" + lc if lc != "en" else ""
            links.append((hrf, "https://cyber-sec-pro.com" + prefix + path))
        links.append(("x-default", "https://cyber-sec-pro.com" + ("/" if path == "" else "/") + path.lstrip("/")))
        return links

    def add_url(loc, lastmod, changefreq, priority, hreflangs=None):
        url = "  <url>\n    <loc>" + loc + "</loc>\n"
        url += "    <lastmod>" + lastmod + "</lastmod>\n"
        url += "    <changefreq>" + changefreq + "</changefreq>\n"
        url += "    <priority>" + str(priority) + "</priority>"
        if hreflangs:
            for hrf, href in hreflangs:
                url += "\n    <xhtml:link rel=\"alternate\" hreflang=\"" + hrf + "\" href=\"" + href + "\" />"
        url += "\n  </url>"
        return url

    urls = []

    # Home pages for all locales
    for lc, hrf in LOCALES:
        prefix = "/" + lc if lc != "en" else ""
        hreflangs = []
        for lc2, hrf2 in LOCALES:
            prefix2 = "/" + lc2 if lc2 != "en" else ""
            hreflangs.append((hrf2, "https://cyber-sec-pro.com" + prefix2 + "/"))
        hreflangs.append(("x-default", "https://cyber-sec-pro.com/"))

        urls.append(add_url(
            "https://cyber-sec-pro.com" + prefix + "/",
            "2026-04-14", "weekly", 1.0 if lc == "en" else 0.9,
            [("en", "https://cyber-sec-pro.com/"), ("tr", "https://cyber-sec-pro.com/tr/"),
             ("de", "https://cyber-sec-pro.com/de/"), ("fr", "https://cyber-sec-pro.com/fr/"),
             ("es", "https://cyber-sec-pro.com/es/"), ("ar", "https://cyber-sec-pro.com/ar/"),
             ("ja", "https://cyber-sec-pro.com/ja/"), ("zh", "https://cyber-sec-pro.com/zh/"),
             ("ru", "https://cyber-sec-pro.com/ru/"), ("ko", "https://cyber-sec-pro.com/ko/"),
             ("x-default", "https://cyber-sec-pro.com/")]
        ))

    # Tools page
    for lc, hrf in LOCALES:
        prefix = "/" + lc if lc != "en" else ""
        hreflangs = []
        for lc2, hrf2 in LOCALES:
            prefix2 = "/" + lc2 if lc2 != "en" else ""
            hreflangs.append((hrf2, "https://cyber-sec-pro.com" + prefix2 + "/tools/"))
        hreflangs.append(("x-default", "https://cyber-sec-pro.com/en/tools/"))

        urls.append(add_url(
            "https://cyber-sec-pro.com" + prefix + "/tools/",
            "2026-04-14", "weekly", 0.9,
            [("en", "https://cyber-sec-pro.com/tools/"), ("tr", "https://cyber-sec-pro.com/tr/tools/"),
             ("de", "https://cyber-sec-pro.com/de/tools/"), ("fr", "https://cyber-sec-pro.com/fr/tools/"),
             ("es", "https://cyber-sec-pro.com/es/tools/"), ("ar", "https://cyber-sec-pro.com/ar/tools/"),
             ("ja", "https://cyber-sec-pro.com/ja/tools/"), ("zh", "https://cyber-sec-pro.com/zh/tools/"),
             ("ru", "https://cyber-sec-pro.com/ru/tools/"), ("ko", "https://cyber-sec-pro.com/ko/tools/"),
             ("x-default", "https://cyber-sec-pro.com/en/tools/")]
        ))

    # Blog posts
    BLOG_POSTS = [
        ("blog/mastering-wireshark", "2026-01-15", "monthly", 0.8),
        ("blog/hashcat-vs-john", "2026-01-12", "monthly", 0.8),
        ("blog/owasp-top-10-2026", "2026-01-08", "monthly", 0.8),
        ("blog/metasploit-zero-to-exploit", "2026-01-05", "monthly", 0.8),
        ("blog/ci-cd-pentest-automation", "2026-01-03", "monthly", 0.8),
        ("blog/wireless-security-assessment", "2025-12-15", "monthly", 0.8),
        ("blog/nmap-network-scanning", "2025-12-10", "monthly", 0.8),
        ("blog/metasploit-exploitation", "2025-12-08", "monthly", 0.8),
        ("blog/hashcat-password-cracking", "2025-11-28", "monthly", 0.8),
    ]

    urls = []

    for slug, lastmod, cf, prio in BLOG_POSTS:
        urls.append("  <url>\n    <loc>https://cyber-sec-pro.com/en/" + slug + "/</loc>\n    <lastmod>" + lastmod + "</lastmod>\n    <changefreq>" + cf + "</changefreq>\n    <priority>" + str(prio) + "</priority>\n    <xhtml:link rel=\"alternate\" hreflang=\"en\" href=\"https://cyber-sec-pro.com/en/" + slug + "/\" />\n    <xhtml:link rel=\"alternate\" hreflang=\"tr\" href=\"https://cyber-sec-pro.com/tr/" + slug + "/\" />\n    <xhtml:link rel=\"alternate\" hreflang=\"x-default\" href=\"https://cyber-sec-pro.com/en/" + slug + "/\" />\n  </url>")

        # Turkish version
        urls.append("  <url>\n    <loc>https://cyber-sec-pro.com/tr/" + slug + "/</loc>\n    <lastmod>" + lastmod + "</lastmod>\n    <changefreq>" + cf + "</changefreq>\n    <priority>" + str(prio * 0.875) + "</priority>\n  </url>")

    # Other pages
    OTHER_PAGES = [
        ("about", "monthly", 0.8),
        ("docs", "weekly", 0.8),
        ("api-reference", "monthly", 0.7),
        ("contact", "monthly", 0.6),
        ("careers", "monthly", 0.6),
        ("security", "monthly", 0.7),
        ("tests", "monthly", 0.7),
        ("privacy", "yearly", 0.3),
        ("terms", "yearly", 0.3),
    ]

    for page, cf, prio in OTHER_PAGES:
        for lc, hrf in LOCALES:
            prefix = "/" + lc if lc != "en" else ""
            hreflangs = []
            for lc2, hrf2 in LOCALES:
                prefix2 = "/" + lc2 if lc2 != "en" else ""
                hreflangs.append((hrf2, "https://cyber-sec-pro.com" + prefix2 + "/" + page + "/"))
            hreflangs.append(("x-default", "https://cyber-sec-pro.com/en/" + page + "/"))
            urls.append(add_url(
                "https://cyber-sec-pro.com" + prefix + "/" + page + "/",
                "2026-04-14", cf, prio, hreflangs
            ))

    # Build XML
    xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<urlset xmlns=\"http://www.sitemaps.org/schemas/sitemap/0.9\" xmlns:xhtml=\"http://www.w3.org/1999/xhtml\">\n" + "\n".join(urls) + "\n</urlset>"

    out_dir = "/home/cybersec/cybersec-pro/frontend/out"
    os.makedirs(out_dir, exist_ok=True)
    with open(os.path.join(out_dir, "sitemap.xml"), "w") as f:
        f.write(xml)
    print("Sitemap generated at", out_dir + "/sitemap.xml")
    print("Total URLs:", len(urls))

if __name__ == "__main__":
    main()

File: frontend/scripts/generate-sitemap.mjs
#!/usr/bin/env node
/**
 * Generate sitemap.xml from what the build actually produced.
 *
 * The previous sitemap was a hand-maintained file in public/. It drifted: by
 * the time anyone checked, 969 of its 1918 URLs were 404s — it still listed
 * 181 tool pages per locale when the catalogue had been curated down to 88 —
 * and 193 real pages, including every blog post and every locale home page,
 * were missing from it entirely. A sitemap that is half dead links is worse
 * than no sitemap: it is the first thing a crawler checks, and it is how
 * search and AI crawlers judge whether the rest of the site is worth trusting.
 *
 * So this walks `out/` instead of being written by hand. Every entry is a page
 * that exists, because it was read off disk after the build made it.
 *
 * Each URL also carries `xhtml:link rel="alternate" hreflang` entries for
 * every locale the same page exists in, plus `x-default` pointing at English.
 * That is what tells a crawler the ten locale copies are one page in ten
 * languages rather than ten near-duplicates competing with each other.
 *
 * Runs as `postbuild`, so `npm run build` cannot produce a stale sitemap.
 */
import { readdirSync, statSync, writeFileSync, existsSync } from 'node:fs';
import { join, relative, sep } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = join(fileURLToPath(new URL('.', import.meta.url)), '..');
const OUT = join(ROOT, 'out');
const ORIGIN = process.env.SITE_ORIGIN || 'https://cyber-sec-pro.com';
const LOCALES = ['en', 'tr', 'de', 'fr', 'es', 'ar', 'ja', 'zh', 'ru', 'ko'];
const DEFAULT_LOCALE = 'en';

/** Pages that exist but should never be advertised to a crawler. */
const EXCLUDE = [
  /^404(\/|$)/,
  /^_not-found(\/|$)/,
  /^_next(\/|$)/,
  /(^|\/)admin(\/|$)/,
  /(^|\/)success(\/|$)/, // post-checkout page, not an entry point
];

/** Crawl priority by page shape. Ordering matters — first match wins. */
const PRIORITY = [
  [/^$/, 1.0], // locale home
  [/^tools$/, 0.9],
  [/^(pricing|security|trust-center)$/, 0.9],
  [/^tools\/[^/]+$/, 0.8],
  [/^blog$/, 0.8],
  [/^blog\/[^/]+$/, 0.7],
  [/^(docs|api-reference)$/, 0.7],
  [/^(about|contact|careers)$/, 0.6],
  [/^(privacy|terms)$/, 0.3],
];

const CHANGEFREQ = [
  [/^$/, 'daily'],
  [/^(blog|tools)$/, 'weekly'],
  [/^blog\/[^/]+$/, 'monthly'],
  [/^(privacy|terms)$/, 'yearly'],
];

const pick = (table, path, fallback) => {
  for (const [re, value] of table) if (re.test(path)) return value;
  return fallback;
};

/** Every directory under out/ that holds an index.html, as a URL path. */
function collectPages(dir, acc = []) {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    if (!entry.isDirectory()) continue;
    const full = join(dir, entry.name);
    if (existsSync(join(full, 'index.html'))) {
      acc.push(relative(OUT, full).split(sep).join('/'));
    }
    collectPages(full, acc);
  }
  return acc;
}

if (!existsSync(OUT)) {
  console.error('generate-sitemap: out/ not found — run `next build` first.');
  process.exit(1);
}

const all = collectPages(OUT);
if (existsSync(join(OUT, 'index.html'))) all.push('');

/**
 * Group by the page's path with its locale stripped, so the ten language
 * copies of one page collapse into a single entry carrying ten alternates.
 * A path with no known locale prefix (the bare `/about/` style) is dropped:
 * those are exactly the URLs the old sitemap advertised into a 404.
 */
const groups = new Map();
for (const page of all) {
  const [head, ...rest] = page.split('/');
  if (!LOCALES.includes(head)) continue;
  const key = rest.join('/');
  if (EXCLUDE.some((re) => re.test(key))) continue;
  if (!groups.has(key)) groups.set(key, new Map());
  groups.get(key).set(head, page);
}

const lastmod = new Date().toISOString().slice(0, 10);
const url = (p) => `${ORIGIN}/${p ? p + '/' : ''}`;

const entries = [...groups.entries()]
  .sort(([a], [b]) => a.localeCompare(b))
  .flatMap(([key, byLocale]) => {
    const alternates = LOCALES.filter((l) => byLocale.has(l))
      .map(
        (l) =>
          `    <xhtml:link rel="alternate" hreflang="${l}" href="${url(byLocale.get(l))}"/>`,
      )
      .join('\n');
    const xDefault = byLocale.has(DEFAULT_LOCALE)
      ? `\n    <xhtml:link rel="alternate" hreflang="x-default" href="${url(byLocale.get(DEFAULT_LOCALE))}"/>`
      : '';

    return [...byLocale.values()].sort().map(
      (page) => `  <url>
    <loc>${url(page)}</loc>
    <lastmod>${lastmod}</lastmod>
    <changefreq>${pick(CHANGEFREQ, key, 'monthly')}</changefreq>
    <priority>${pick(PRIORITY, key, 0.5).toFixed(1)}</priority>
${alternates}${xDefault}
  </url>`,
    );
  });

const xml = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:xhtml="http://www.w3.org/1999/xhtml">
${entries.join('\n')}
</urlset>
`;

writeFileSync(join(OUT, 'sitemap.xml'), xml);
console.log(
  `generate-sitemap: ${entries.length} URLs across ${groups.size} pages x ${LOCALES.length} locales -> out/sitemap.xml`,
);
